### PR TITLE
BUG: Fix ma dot to always return masked array

### PR DESCRIPTION
**Pull request by charris**
Refactor ma.dot to fix regression reported in #6611.

The main functionality is now implemented in the ma.dot function. The following changes were required.
- the `ma.dot` function was moved from `ma.extras` to `ma.core`.
- the `mask_rowcols` function was moved from `ma.extras` to `ma.core`.
- the keyword `out` was added to `ma.dot.`
- tests for `out` and #6611 were added.

For backwards compatibility both `dot` and `mask_rowcols` can still be imported from `extras.py` and are included in `extras.__all__`.

### CODE REVIEW

**Comment by charris** [comment]
Note that the `strict` keyword in `ma.dot` is not currently part of the `dot` method. I think it should be, and done so as to make the signatures of both the function and the method compatible. That will not be backwards compatible with current 1.10.1.

File: numpy/ma/core.py [hunk 1]
```diff
@@ -1224,6 +1227,177 @@ def __str__(self):
 fmod = _DomainedBinaryOperation(umath.fmod, _DomainSafeDivide(), 0, 1)
 mod = _DomainedBinaryOperation(umath.mod, _DomainSafeDivide(), 0, 1)
 
+# Include masked dot here as it is part of the __numpy_ufunc__ system. This
```
Comment (ahaldane) [inline numpy/ma/core.py]: I don't think `dot` is part of the ufunc system *yet*, though there are hopes for it (see #4164). 

I'd be slightly inclined to move the definition of `dot` to be just after the definitions of `inner` and `outer` near the bottom of the file.

**Comment by ahaldane** [comment]
Re: the `strict` keyword, it is also discussed in #5739  (which this PR fixes). 

I haven't thought about it terribly deeply, but I like the idea of having `strict` in the method too. #5739 mentions there is a danger if `np.dot` ever gets new keyword args, but hopefully we can just remember to update `ma.dot` at the same time if that happens.

File: numpy/ma/core.py (hunk 1 above)
Comment (charris) [inline numpy/ma/core.py]: `dot`, both method and function, as well as `matmul` (@) check `__numpy_ufunc__`, although none of them are ufuncs ;) I don't have a problem with moving the definitions, I wasn't sure where the best spot was.

**Comment by charris** [comment]
So would you argue to put strict after `out` as suggested in #5739?

**Comment by ahaldane** [comment]
Here are some thoughts on the argument order, after loking some more:

First, I don't feel too bad adding a keywords `ndarrays` don't have, since number of other `MaskedArray` methods already add their own keywords (eg many have `fill_value`).

Furthermore, note that my PR #5706 is going to be adding a lot of missing keyword args (I *do* hope to finish that sometime). It looks like my strategy there was to keep the nonstandard keywords (eg, see [min](https://github.com/numpy/numpy/pull/5706/files#diff-e85f023d9c07df014a190b485db72c3dR5350)). But because of back-compat issues I have the nonstandard args *before* the standard args like `out`. It seems the order of the args isn't going to be terribly consistent.

Maybe a solution is to accept `strict` though `**kwargs`. That way no one will be tempted to use it as a positional arg, and then get tripped up later if we add new keywords. That is have something like

    def dot(self, other, out=None, **kwds):
        strict = kwds.pop('strict', False)
        if len(kwds) != 0:
            raise TypeError("ma.dot got an unexpected keyword argument")

Does that work?

**Comment by charris** [comment]
Moved `dot` and `mask_rowcols`. I don't think the keyword fix works for the `dot` function as `strict` already has the third position. @abalkin's suggestion doesn't work either, it is a syntax violation. What does work is to have only two positional arguments, but that is a bit ugly. The easiest thing might be to just add `out` to the end after `strict` and check that `strict` is a boolean. It's incompatible with non masked array `dot`, but allows the two positional transition at a later time if we want to go that way.

**Comment by ahaldane** [comment]
That sounds reasonable.

If there might be opportunity to flip the keyword order sometime in the future, then I agree adding `out` last in `ma.dot` seems like the easy conservative change for 1.10.2.

For `MaskedArray.dot`, I would be fine with either adding `strict` last or adding `**kwds` to handle strict, but maybe we can leave that for 1.11 since it isn't causing any bugs right now.

In that case the only thing left is maybe adding a check that `strict` is a boolean in `ma.dot`?

**Comment by charris** [comment]
@ahaldane I decided to skip the boolean check as some folks may use ints or other objects that have boolean values, so a bit risky at this point. Can be added later. Thanks for the review.

**Pull request merged**